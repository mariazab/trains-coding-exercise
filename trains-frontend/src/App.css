.App {
  text-align: center;
  padding: 1%;
}

.main-header {
  margin: 3%;
}

.login-header {
  margin: 3% 0 2%;
}

.alert {
  margin: 3% 35%;
}

.form {
  padding: 1%;
}

.link {
  text-decoration: none;
  color: white;
}

.logout-btn {
  margin: 5%;
}

.content {
  margin: 3%;
}