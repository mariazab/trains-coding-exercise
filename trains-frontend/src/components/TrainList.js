import React, { Component } from "react";
import {SERVER_URL} from "../constants.js";
import { TableContainer, TableHead, TableRow, TableCell, Table,
        Paper, Typography, Button, 
        Dialog, DialogTitle, DialogContent, TextField, DialogActions } from "@material-ui/core";

//Component with the list of trains
class TrainList extends Component {

    constructor(props) {
        super(props);
        this.state = { trains: [], openDialog: false, lat: "", lon: "", currentTrainId: "", currentTrain: {} };
    }

    componentDidMount() {
        this.fetchTrains();
    }

    //Fetching trains with the token and saving it to state
    fetchTrains = () => {
        //Get the token from the session storage
        const token = sessionStorage.getItem("jwtToken");
        fetch(SERVER_URL + "trains", {
            headers: {"Authorization": token}
        })
        .then((response) => response.json())
        .then((responseData) => {
            this.setState({
                trains: responseData,
            });
        })
        .catch(err => console.error(err));
    }

    //When the 'Update Location' button is clicked, fetch the data of this train and open the dialog
    openUpdateDialog = (id) => {
        const token = sessionStorage.getItem("jwtToken");
        fetch(SERVER_URL + "trains/" + id, {
            headers: {"Authorization": token}
        })
        .then((response) => response.json())
        .then((responseData) => {
            console.log(responseData);
            this.setState({
                currentTrain: responseData,
                openDialog: true,
            });
        })
        .catch(err => console.error(err));
    }

    //Closing the dialog
    handleClose = () => {
        this.setState({openDialog: false});
    }

    //Saving the data from the fields from the dialog
    handleChange = (event) => {
        this.setState({ [event.target.name]: event.target.value});
    }

    //Update the train's location with HTTP PUT request
    update = () => {

        let currentTrainId = this.state.currentTrain.id;

        //Get the current values of the train
        let train = {...this.state.currentTrain};

        //Parse the Lat and Lon strings, if the strings are empty, this will return NaN
        let newLat = parseFloat((this.state.lat).replace(",", "."));
        let newLon = parseFloat((this.state.lon).replace(",", "."));

        //If the Lat and Lon are both Nan, return and go back to train list without changing anything
        if(isNaN(newLat) && isNaN(newLon)) {
            this.handleClose();
            return;
        }
        //If Lat is NaN, update only the Lon, the Lat stays the same
        else if (isNaN(newLat)) {
            train.coordinates[1] = newLon;
        } 
        //If the Lon is NaN, update only the Lat, the Lon stays the same
        else if(isNaN(newLon)) {
            train.coordinates[0] = newLat;
        } 
        //Otherwise, update both values
        else {
            train.coordinates = [newLat, newLon];
        }

        //Send the train to the server and update the state with the response
        const token = sessionStorage.getItem("jwtToken");
        fetch(SERVER_URL + "trains/" + currentTrainId + "/location", {
            method: "PUT", 
            headers: {"Authorization": token, "Content-Type": "application/json"},
            body: JSON.stringify(train)
        })
        .then((response) => response.json())
        .then((responseData) => {
            this.setState({
                trains: responseData,
            })

        })
        .catch(err =>console.error(err));
        
        //Close the dialog
        this.handleClose();

        //Reset the dialog's fields
        this.setState({lat: "", lon: ""});
    }

    render() {
        const trainRows = this.state.trains.map((train, index) =>
            <TableRow key={index}>
                <TableCell>{train.name}</TableCell>
                <TableCell>{train.destination}</TableCell>
                <TableCell>{train.speed}</TableCell>
                <TableCell>Latitude: {train.coordinates[0]}
                            <br />
                            Longitude: {train.coordinates[1]}
                </TableCell>
                <TableCell>
                <Button variant="outlined" color="secondary" size="small" onClick={() => this.openUpdateDialog(train.id)}>Update location</Button>
                </TableCell>
            </TableRow>
        );

        return (
            <div className="content">

            {/* Dialog with fields to update train */}
            <Dialog open={this.state.openDialog} onClose={this.handleClose}>
                <DialogTitle>Update location of {this.state.currentTrain.name} train to {this.state.currentTrain.destination}</DialogTitle>
                <DialogContent>
                    <TextField autoFocus
                                name="lat" label="Latitude" 
                                variant="outlined" onChange={this.handleChange} value={this.state.lat}/>
                    <TextField name="lon" label="Longitude"
                                variant="outlined" onChange={this.handleChange} value={this.state.lon}/>
                </DialogContent>
                <DialogActions>
                    <Button onClick={this.update}>Save</Button>
                    <Button onClick={this.handleClose}>Cancel</Button>
                </DialogActions>
            </Dialog>

            <TableContainer component={Paper}>
                <Table>
                    <TableHead>
                        <TableRow>
                            <TableCell><Typography variant="h6">Train Name</Typography></TableCell>
                            <TableCell><Typography variant="h6">Destination</Typography></TableCell>
                            <TableCell><Typography variant="h6">Speed</Typography></TableCell>
                            <TableCell><Typography variant="h6">Location</Typography></TableCell>
                            <TableCell></TableCell>
                        </TableRow>
                            {trainRows}
                    </TableHead>
                </Table>
            </TableContainer>
            </div>
        );
    }
}

export default TrainList;