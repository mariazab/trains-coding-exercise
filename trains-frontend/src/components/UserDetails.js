import React, { Component } from "react";
import {SERVER_URL} from "../constants.js";
import "../App.css";
import Typography from "@material-ui/core/Typography";

//Component with details of the user
class UserDetails extends Component {

    constructor(props) {
        super(props);
        this.state = { user: {} };
    }

    componentDidMount() {
        this.fetchUser();
    }

    //Fetch the user information and save it to state
    fetchUser = () => {
        //Get the token and username from the session storage
        const token = sessionStorage.getItem("jwtToken");
        const username = sessionStorage.getItem("username");
        fetch(SERVER_URL + "users/" + username, {
            headers: {"Authorization": token}
        })
        .then((response) => response.json())
        .then((responseData) => {
            this.setState({
                user: responseData,
            });
    })
        
        .catch(err => console.error(err));
    }

    render() {

        return (
            <div className="content"> 
                <Typography variant="h4">User details</Typography>
                <br /><br />
                <Typography variant="h6">Username: {this.state.user.username}</Typography>
                <Typography variant="h6">Full name: {this.state.user.fullName}</Typography>
                <Typography variant="h6">Email: {this.state.user.email}</Typography>
                
            </div>
        );
    }
}

export default UserDetails;