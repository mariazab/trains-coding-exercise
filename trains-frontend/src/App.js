import React from "react";
import "./App.css";
import Typography from "@material-ui/core/Typography";
import Login from "./components/Login";

function App() {
  return (
    <div className="App">
      <div className="main-header">
          <Typography variant="h3">Location of trains</Typography>
        </div>
      <Login />
    </div>
  );
}

export default App;