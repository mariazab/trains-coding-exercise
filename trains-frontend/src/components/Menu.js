import React, { useState } from "react";
import "../App.css";
import { BrowserRouter, Switch, Route, Link } from "react-router-dom";
import Button from "@material-ui/core/Button";
import ButtonGroup from "@material-ui/core/ButtonGroup";
import TrainList from "./TrainList";
import UserDetails from "./UserDetails";
import Login from "./Login";

//Menu after the user is logged in, links to TrainList and UserDetails and logout button
const Menu = () => {

    const [isLoggedIn, setLoginState] = useState(true);

    const logout = () => {
        sessionStorage.removeItem("jwtToken");
        sessionStorage.removeItem("usernmame");
        setLoginState(false);
    }

    if (isLoggedIn === false) {
        return <Login />;
    }
    else {
    return (
        <div>       
            <BrowserRouter>
                <div>
                    <ButtonGroup variant="contained" color="primary" size="medium">
                        <Button><Link className="link" to="/">Train List</Link></Button>
                        <Button><Link className="link" to="/user">User details</Link></Button>
                    </ButtonGroup>
                    <Switch>
                        <Route exact path="/" component={TrainList} />
                        <Route path="/user" component={UserDetails} />
                    </Switch>
                </div>
            </BrowserRouter>
            
            <div className="logout-btn">
                <Button variant="outlined" color="secondary" size="medium" onClick={logout}>Logout</Button>
            </div>
        </div>
    )}
}

export default Menu;