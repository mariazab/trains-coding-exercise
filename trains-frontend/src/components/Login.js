import React, { useState } from "react";
import {SERVER_URL} from "../constants";
import "../App.css";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import { Alert, AlertTitle } from "@material-ui/lab";
import Typography from "@material-ui/core/Typography";
import Menu from "./Menu";
import Register from "./Register";

//Login page
const Login = () => {

    const [user, setUser] = useState({username: "", password: ""});
    const [isAuthenticated, setAuth] = useState(false);
    const [failedLogin, setFailedLogin] = useState(false);
    const [isRegistering, setRegistration] = useState(false);

    //Save values from login fields
    const handleChange = (event) => {
        setUser({...user, [event.target.name]: event.target.value})
    }

    //Login the user and save the received token
    const login = () => {
        fetch(SERVER_URL + "login", {
            method: "POST",
            body: JSON.stringify(user)
        })
        .then(res => {
            const jwtToken = res.headers.get("Authorization");
            if(jwtToken !== null) {
                sessionStorage.setItem("jwtToken", jwtToken);
                sessionStorage.setItem("username", user.username);
                setAuth(true);
            }
            else {
                setFailedLogin(true);
            }
        })
        .catch(err => console.error(err))
    }

    //When Register button is clicked, go to registration
    const register = () => {
        setRegistration(true);
    }

    //Render different components
    //If login is successfull, go to '/' path with TrainList and Menu
    if (isAuthenticated === true) {
        return (<Menu />);
    }
    else if(isRegistering === true) {
        return <Register />;
    }
    else {
        //Alert div for unsuccessful login
        let alert;
        if (failedLogin) {
            alert = <div className="alert"><Alert severity="error"><AlertTitle>Invalid username or password.</AlertTitle></Alert></div>;
        }
        return (
            <div>
                <div className="login-header">
                    <Typography variant="h4">Welcome!</Typography>
                </div>
                
                {alert}
                
                <div className="form">
                    <TextField name="username" label="Username" variant="outlined" className="login-form" onChange={handleChange} />
                </div>
                
                <div className="form"> 
                    <TextField type="password" name="password" label="Password" variant="outlined" className="login-form" onChange={handleChange} />
                </div>
                
                <div className="form">
                    <Button variant="contained" color="primary" size="large" onClick={login}>Login</Button>
                </div>
                
                <div className="form">
                    <Button variant="outlined" color="primary" size="small" onClick={register}>Register</Button>
                </div>
            </div>
        );
    }
}

export default Login;