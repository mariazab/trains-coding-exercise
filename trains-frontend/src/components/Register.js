import React, { Component } from "react";
import {SERVER_URL} from "../constants";
import "../App.css";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";
import Dialog from "@material-ui/core/Dialog";
import DialogTitle from "@material-ui/core/DialogTitle";
import DialogActions from "@material-ui/core/DialogActions";
import Login from "./Login";

//Component for Registration
class Register extends Component {

    constructor(props) {
        super(props);
        this.state = { username: "", password: "", passwordCheck: "", fullName: "", email: "", 
                        userAlreadyExists: false, succeeded: false, openDialog: false, dialogText: ""};
    }

    //Save values from the fields to state
    handleChange = (event) => {
        this.setState({ [event.target.name]: event.target.value});
    }

    //Check if any of the fields were empty
    checkIfEmtpyFields = () => {
        let username = (this.state.username).trim(); 
        let password = (this.state.password).trim();
        let fullName = (this.state.fullName).trim();
        let email = (this.state.email).trim();
        
        if(username === "" || password === "" || fullName === "" | email === "") {
            return true;
        }
    
        return false;
    }

    //Check if the password and passwordCheck match
    checkPassword = () => {
        let password = this.state.password;
        let passwordCheck = this.state.passwordCheck;

        if(password === passwordCheck) {
            return true;
        }
        else {
            return false;
        }
    }

    //Create a hash from the password
    hashPassword = () => {
        let password = this.state.password;
        let passwordHash = "";
        let bcrypt = require("bcryptjs");

        let salt = bcrypt.genSaltSync(12);
        passwordHash = bcrypt.hashSync(password, salt);

        return passwordHash;
    }

    //Send the new user to the server
    register = () => {

        //First check if passwords match and there are no empty fields
        let passwordMatch = this.checkPassword();
        let emptyFields = this.checkIfEmtpyFields();

        //If so, proceed with the registration
        if(passwordMatch === true && emptyFields === false) {

            let passwordHash = this.hashPassword();

            const user = {username: (this.state.username).trim(), password: passwordHash, fullName: (this.state.fullName).trim(), email: (this.state.email).trim(), role: "user", loadLogin: false};

            fetch(SERVER_URL + "api/users", {
                method: "POST", 
                headers: {"Content-Type": "application/json"},
                body: JSON.stringify(user)
            })
            .then(res => {
                //Based on different response status, set the dialog message
                //Success
                if (res.status === 201) {
                   this.setState({openDialog: true, dialogText: "Registration successful!", succeeded: true});                    
                }
                //Conflict of PK username
                else if (res.status === 409) {
                    this.setState({openDialog: true, dialogText: "This username already exists."});
                }
                //Failure
                else {
                    this.setState({openDialog: true, dialogText: "Registration failed, try again"});
                }
    
            })
            .catch(err => console.error(err))

        }
        //If passwords don't match, prompt user to try again
        else if (passwordMatch === false){
            this.setState({openDialog: true, dialogText: "Passwords don't match, try again."});
        }
        //If any fields are empty, prompt user to try again
        else if(emptyFields === true) {
            this.setState({openDialog: true, dialogText: "None of the fields can be empty, try again."});
        }

    }

    //Handling dialog close, if registration succeeded, go to Login
    handleClose = () => {
        this.setState({openDialog: false});
        if (this.state.succeeded === true) {
            this.setState({loadLogin: true});
        }
    }

    //If 'Go to Login' button is clicked, load Login
    goToLoginBtn = () => {
        this.setState({loadLogin: true});
    }

    render() {

        if (this.state.loadLogin === true) {
            return <Login />;
        }
        else {
            return (
                <div>
                    {/* Dialog with the message if the registration was successful or not */}
                    <Dialog open={this.state.openDialog} onClose={this.handleClose}>
                        <DialogTitle id="title">{this.state.dialogText}</DialogTitle>
                        <DialogActions>
                            <Button onClick={this.handleClose} color="primary">Close</Button>
                        </DialogActions>
                    </Dialog>
                    
                    <div className="login-header">
                        <Typography variant="h5">Create an account</Typography>
                    </div>
                
                    <div className="form">
                        <TextField required 
                                name="username" label="Username" 
                                variant="outlined" onChange={this.handleChange} value={this.state.username} />
                    </div>
                    
                    <div className="form">
                        <TextField required 
                                type="password" name="password" 
                                label="Password" variant="outlined" 
                                onChange={this.handleChange} value={this.state.password} />
                    </div>
                    
                    <div className="form">
                        <TextField required 
                                type="password" name="passwordCheck" 
                                label="Confirm password" variant="outlined" 
                                onChange={this.handleChange} value={this.state.passwordCheck} />
                    </div>
                    
                    <div className="form">
                        <TextField required 
                                name="fullName" label="Full Name" 
                                variant="outlined" onChange={this.handleChange} value={this.state.fullName} />
                    </div>
                    
                    <div className="form">
                        <TextField required 
                                type="email" name="email" 
                                label="Email" variant="outlined" 
                                onChange={this.handleChange} value={this.state.email} />
                    </div>
                    
                    <div className="form">
                    <Button variant="contained" color="primary" size="large" onClick={this.register}>Register</Button>
                    </div>

                    <div className="form">
                    <Button variant="outlined" color="primary" size="small" onClick={this.goToLoginBtn}>Go to Login</Button>
                    </div>
                </div>
            )
        }
    }   
    
}

export default Register;