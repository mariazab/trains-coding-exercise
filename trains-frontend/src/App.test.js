import React from "react"; 
import ReactDOM from "react-dom";
import renderer from "react-test-renderer"; 
import App from "./App";
import Menu from "./components/Menu";
import Login from "./components/Login";
import Register from "./components/Register";
import TrainList from "./components/TrainList";
import UserDetails from "./components/UserDetails";

it("renders without crashing", () => {  
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it("Menu renders correctly", () => {
  const tree = renderer.create(<Menu/>).toJSON;
  expect(tree).toMatchSnapshot();
});

it("Login renders correctly", () => {
  const tree = renderer.create(<Login/>).toJSON;
  expect(tree).toMatchSnapshot();
});

it("Register renders correctly", () => {
  const tree = renderer.create(<Register/>).toJSON;
  expect(tree).toMatchSnapshot();
});

it("TrainList renders correctly", () => {
  const tree = renderer.create(<TrainList/>).toJSON;
  expect(tree).toMatchSnapshot();
});

it("UserDetails renders correctly", () => {
  const tree = renderer.create(<UserDetails/>).toJSON;
  expect(tree).toMatchSnapshot();
});