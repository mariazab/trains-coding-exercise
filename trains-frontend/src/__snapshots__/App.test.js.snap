// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctly 1`] = `[Function]`;

exports[`Menu renders correctly 1`] = `[Function]`;

exports[`Register renders correctly 1`] = `[Function]`;

exports[`TrainList renders correctly 1`] = `[Function]`;

exports[`UserDetails renders correctly 1`] = `[Function]`;
